File: src/stores/downloads.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'

export type DownloadStatus = 'pending' | 'downloading' | 'completed' | 'failed' | 'cancelled' | 'paused'

export interface Download {
  id: string
  url: string
  filename: string
  file_path: string
  total_bytes: number
  downloaded_bytes: number
  status: DownloadStatus
  start_time: string
  end_time?: string
  error_message?: string
  mime_type?: string
  referrer?: string
}

export interface DownloadStats {
  total_downloads: number
  completed_downloads: number
  failed_downloads: number
  total_bytes_downloaded: number
  active_downloads: number
}

export const useDownloadsStore = defineStore('downloads', () => {
  const downloads = ref<Download[]>([])
  const isLoading = ref(false)

  async function loadDownloads(): Promise<void> {
    try {
      isLoading.value = true
      const downloadList = await invoke<Download[]>('get_downloads')
      downloads.value = downloadList
    } catch (error) {
      console.error('Failed to load downloads:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  async function startDownload(url: string, filename?: string, referrer?: string): Promise<string> {
    try {
      const downloadId = await invoke<string>('start_download', {
        url,
        filename,
        referrer
      })
      
      await loadDownloads()
      return downloadId
    } catch (error) {
      console.error('Failed to start download:', error)
      throw error
    }
  }

  async function cancelDownload(downloadId: string): Promise<void> {
    try {
      await invoke('cancel_download', { downloadId })
      
      const download = downloads.value.find(d => d.id === downloadId)
      if (download) {
        download.status = 'cancelled'
      }
    } catch (error) {
      console.error('Failed to cancel download:', error)
      throw error
    }
  }

  async function pauseDownload(downloadId: string): Promise<void> {
    try {
      await invoke('pause_download', { downloadId })
      
      const download = downloads.value.find(d => d.id === downloadId)
      if (download) {
        download.status = 'paused'
      }
    } catch (error) {
      console.error('Failed to pause download:', error)
      throw error
    }
  }

  async function resumeDownload(downloadId: string): Promise<void> {
    try {
      await invoke('resume_download', { downloadId })
      
      const download = downloads.value.find(d => d.id === downloadId)
      if (download) {
        download.status = 'downloading'
      }
    } catch (error) {
      console.error('Failed to resume download:', error)
      throw error
    }
  }

  async function removeDownload(downloadId: string): Promise<void> {
    try {
      await invoke('remove_download', { downloadId })
      downloads.value = downloads.value.filter(d => d.id !== downloadId)
    } catch (error) {
      console.error('Failed to remove download:', error)
      throw error
    }
  }

  async function clearCompletedDownloads(): Promise<void> {
    try {
      await invoke('clear_completed_downloads')
      downloads.value = downloads.value.filter(d => d.status !== 'completed')
    } catch (error) {
      console.error('Failed to clear completed downloads:', error)
      throw error
    }
  }

  async function getActiveDownloads(): Promise<Download[]> {
    try {
      const activeDownloads = await invoke<Download[]>('get_active_downloads')
      return activeDownloads
    } catch (error) {
      console.error('Failed to get active downloads:', error)
      return []
    }
  }

  async function getDownloadStats(): Promise<DownloadStats> {
    try {
      const stats = await invoke<DownloadStats>('get_download_stats')
      return stats
    } catch (error) {
      console.error('Failed to get download stats:', error)
      return {
        total_downloads: 0,
        completed_downloads: 0,
        failed_downloads: 0,
        total_bytes_downloaded: 0,
        active_downloads: 0
      }
    }
  }

  async function setDownloadDirectory(directory: string): Promise<void> {
    try {
      await invoke('set_download_directory', { directory })
    } catch (error) {
      console.error('Failed to set download directory:', error)
      throw error
    }
  }

  async function fetchDownloadProgress(downloadId: string): Promise<number> {
    try {
      const progress = await invoke<number>('get_download_progress', { downloadId })
      return progress
    } catch (error) {
      console.error('Failed to get download progress:', error)
      return 0
    }
  }

  async function exportDownloads(): Promise<string> {
    try {
      const data = await invoke<string>('export_downloads')
      return data
    } catch (error) {
      console.error('Failed to export downloads:', error)
      throw error
    }
  }

  function getDownloadById(downloadId: string): Download | undefined {
    return downloads.value.find(download => download.id === downloadId)
  }

  function getDownloadsByStatus(status: DownloadStatus): Download[] {
    return downloads.value.filter(download => download.status === status)
  }

  function getCompletedDownloads(): Download[] {
    return downloads.value.filter(download => download.status === 'completed')
  }

  function getFailedDownloads(): Download[] {
    return downloads.value.filter(download => download.status === 'failed')
  }

  function calculateDownloadProgress(download: Download): number {
    if (download.total_bytes === 0) return 0
    return Math.round((download.downloaded_bytes / download.total_bytes) * 100)
  }

  function formatFileSize(bytes: number): string {
    const units = ['B', 'KB', 'MB', 'GB', 'TB']
    let size = bytes
    let unitIndex = 0
    
    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024
      unitIndex++
    }
    
    return `${size.toFixed(1)} ${units[unitIndex]}`
  }

  function getDownloadSpeed(download: Download): string {
    if (download.status !== 'downloading' || !download.start_time) {
      return '0 B/s'
    }
    
    const startTime = new Date(download.start_time).getTime()
    const currentTime = Date.now()
    const elapsedSeconds = (currentTime - startTime) / 1000
    
    if (elapsedSeconds === 0) return '0 B/s'
    
    const bytesPerSecond = download.downloaded_bytes / elapsedSeconds
    return `${formatFileSize(bytesPerSecond)}/s`
  }

  function getEstimatedTimeRemaining(download: Download): string {
    if (download.status !== 'downloading' || download.total_bytes === 0) {
      return 'Unknown'
    }
    
    const remainingBytes = download.total_bytes - download.downloaded_bytes
    const startTime = new Date(download.start_time).getTime()
    const currentTime = Date.now()
    const elapsedSeconds = (currentTime - startTime) / 1000
    
    if (elapsedSeconds === 0 || download.downloaded_bytes === 0) {
      return 'Unknown'
    }
    
    const bytesPerSecond = download.downloaded_bytes / elapsedSeconds
    const remainingSeconds = remainingBytes / bytesPerSecond
    
    if (remainingSeconds < 60) {
      return `${Math.round(remainingSeconds)}s`
    } else if (remainingSeconds < 3600) {
      return `${Math.round(remainingSeconds / 60)}m`
    } else {
      return `${Math.round(remainingSeconds / 3600)}h`
    }
  }

  return {
    downloads,
    isLoading,
    loadDownloads,
    startDownload,
    cancelDownload,
    pauseDownload,
    resumeDownload,
    removeDownload,
    clearCompletedDownloads,
    getActiveDownloads,
    getDownloadStats,
    setDownloadDirectory,
    fetchDownloadProgress,
    exportDownloads,
    getDownloadById,
    getDownloadsByStatus,
    getCompletedDownloads,
    getFailedDownloads,
    formatFileSize,
    calculateDownloadProgress,
    getDownloadSpeed,
    getEstimatedTimeRemaining
  }
})

File: src-tauri/src/browser/downloads.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use uuid::Uuid;
use tokio::sync::RwLock;
use once_cell::sync::Lazy;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum DownloadStatus {
    Pending,
    InProgress,
    Completed,
    Failed,
    Cancelled,
    Paused,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Download {
    pub id: String,
    pub url: String,
    pub filename: String,
    pub file_path: PathBuf,
    pub mime_type: Option<String>,
    pub total_bytes: Option<u64>,
    pub downloaded_bytes: u64,
    pub status: DownloadStatus,
    pub start_time: chrono::DateTime<chrono::Utc>,
    pub end_time: Option<chrono::DateTime<chrono::Utc>>,
    pub error_message: Option<String>,
    pub referrer: Option<String>,
    pub user_agent: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadStats {
    pub total_downloads: u64,
    pub completed_downloads: u64,
    pub failed_downloads: u64,
    pub total_bytes_downloaded: u64,
    pub active_downloads: u64,
}

static DOWNLOAD_MANAGER: Lazy<RwLock<DownloadManager>> = Lazy::new(|| {
    RwLock::new(DownloadManager::new())
});

pub struct DownloadManager {
    pub downloads: HashMap<String, Download>,
    pub download_directory: PathBuf,
}

impl DownloadManager {
    pub fn new() -> Self {
        let download_directory = dirs::download_dir()
            .unwrap_or_else(|| std::env::current_dir().unwrap_or_default());
        
        Self {
            downloads: HashMap::new(),
            download_directory,
        }
    }

    pub fn start_download(
        &mut self,
        url: &str,
        filename: Option<&str>,
        referrer: Option<&str>,
    ) -> Result<String, String> {
        let download_id = Uuid::new_v4().to_string();
        
        let filename = filename
            .map(|f| f.to_string())
            .or_else(|| self.extract_filename_from_url(url))
            .unwrap_or_else(|| format!("download_{}", download_id));
        
        let file_path = self.download_directory.join(&filename);
        
        let download = Download {
            id: download_id.clone(),
            url: url.to_string(),
            filename,
            file_path,
            mime_type: None,
            total_bytes: None,
            downloaded_bytes: 0,
            status: DownloadStatus::Pending,
            start_time: chrono::Utc::now(),
            end_time: None,
            error_message: None,
            referrer: referrer.map(|r| r.to_string()),
            user_agent: Some("Sw3do Browser/1.0".to_string()),
        };
        
        self.downloads.insert(download_id.clone(), download);
        Ok(download_id)
    }

    pub fn update_download_progress(
        &mut self,
        download_id: &str,
        downloaded_bytes: u64,
        total_bytes: Option<u64>,
    ) -> Result<(), String> {
        let download = self.downloads.get_mut(download_id)
            .ok_or("Download not found")?;
        
        download.downloaded_bytes = downloaded_bytes;
        if let Some(total) = total_bytes {
            download.total_bytes = Some(total);
        }
        download.status = DownloadStatus::InProgress;
        
        Ok(())
    }

    pub fn complete_download(&mut self, download_id: &str) -> Result<(), String> {
        let download = self.downloads.get_mut(download_id)
            .ok_or("Download not found")?;
        
        download.status = DownloadStatus::Completed;
        download.end_time = Some(chrono::Utc::now());
        
        Ok(())
    }

    pub fn fail_download(&mut self, download_id: &str, error: &str) -> Result<(), String> {
        let download = self.downloads.get_mut(download_id)
            .ok_or("Download not found")?;
        
        download.status = DownloadStatus::Failed;
        download.error_message = Some(error.to_string());
        download.end_time = Some(chrono::Utc::now());
        
        Ok(())
    }

    pub fn cancel_download(&mut self, download_id: &str) -> Result<(), String> {
        let download = self.downloads.get_mut(download_id)
            .ok_or("Download not found")?;
        
        download.status = DownloadStatus::Cancelled;
        download.end_time = Some(chrono::Utc::now());
        
        Ok(())
    }

    pub fn pause_download(&mut self, download_id: &str) -> Result<(), String> {
        let download = self.downloads.get_mut(download_id)
            .ok_or("Download not found")?;
        
        if matches!(download.status, DownloadStatus::InProgress) {
            download.status = DownloadStatus::Paused;
        }
        
        Ok(())
    }

    pub fn resume_download(&mut self, download_id: &str) -> Result<(), String> {
        let download = self.downloads.get_mut(download_id)
            .ok_or("Download not found")?;
        
        if matches!(download.status, DownloadStatus::Paused) {
            download.status = DownloadStatus::InProgress;
        }
        
        Ok(())
    }

    pub fn remove_download(&mut self, download_id: &str) -> Result<(), String> {
        self.downloads.remove(download_id)
            .ok_or("Download not found")?;
        Ok(())
    }

    pub fn clear_completed_downloads(&mut self) {
        self.downloads.retain(|_, download| {
            !matches!(download.status, DownloadStatus::Completed | DownloadStatus::Failed | DownloadStatus::Cancelled)
        });
    }

    pub fn get_downloads(&self) -> Vec<&Download> {
        let mut downloads: Vec<&Download> = self.downloads.values().collect();
        downloads.sort_by(|a, b| b.start_time.cmp(&a.start_time));
        downloads
    }

    pub fn get_active_downloads(&self) -> Vec<&Download> {
        self.downloads.values()
            .filter(|download| matches!(download.status, DownloadStatus::InProgress | DownloadStatus::Paused))
            .collect()
    }

    pub fn get_download_stats(&self) -> DownloadStats {
        let total_downloads = self.downloads.len() as u64;
        let completed_downloads = self.downloads.values()
            .filter(|d| matches!(d.status, DownloadStatus::Completed))
            .count() as u64;
        let failed_downloads = self.downloads.values()
            .filter(|d| matches!(d.status, DownloadStatus::Failed))
            .count() as u64;
        let total_bytes_downloaded = self.downloads.values()
            .map(|d| d.downloaded_bytes)
            .sum();
        let active_downloads = self.downloads.values()
            .filter(|d| matches!(d.status, DownloadStatus::InProgress | DownloadStatus::Paused))
            .count() as u64;
        
        DownloadStats {
            total_downloads,
            completed_downloads,
            failed_downloads,
            total_bytes_downloaded,
            active_downloads,
        }
    }

    pub fn set_download_directory(&mut self, path: PathBuf) -> Result<(), String> {
        if !path.exists() {
            std::fs::create_dir_all(&path)
                .map_err(|e| format!("Failed to create download directory: {}", e))?;
        }
        
        if !path.is_dir() {
            return Err("Path is not a directory".to_string());
        }
        
        self.download_directory = path;
        Ok(())
    }

    fn extract_filename_from_url(&self, url: &str) -> Option<String> {
        if let Ok(parsed_url) = url::Url::parse(url) {
            if let Some(segments) = parsed_url.path_segments() {
                if let Some(last_segment) = segments.last() {
                    if !last_segment.is_empty() {
                        return Some(last_segment.to_string());
                    }
                }
            }
        }
        None
    }

    pub fn get_download_progress(&self, download_id: &str) -> Option<f64> {
        let download = self.downloads.get(download_id)?;
        
        if let Some(total_bytes) = download.total_bytes {
            if total_bytes > 0 {
                return Some((download.downloaded_bytes as f64 / total_bytes as f64) * 100.0);
            }
        }
        
        None
    }

    pub fn export_downloads(&self) -> Result<String, String> {
        let downloads: Vec<&Download> = self.downloads.values().collect();
        serde_json::to_string_pretty(&downloads)
            .map_err(|e| format!("Failed to export downloads: {}", e))
    }
}

#[tauri::command]
pub async fn start_download(url: String, filename: Option<String>, referrer: Option<String>) -> Result<String, String> {
    let mut manager = DOWNLOAD_MANAGER.write().await;
    manager.start_download(&url, filename.as_deref(), referrer.as_deref())
}

#[tauri::command]
pub async fn cancel_download(download_id: String) -> Result<(), String> {
    let mut manager = DOWNLOAD_MANAGER.write().await;
    manager.cancel_download(&download_id)
}

#[tauri::command]
pub async fn pause_download(download_id: String) -> Result<(), String> {
    let mut manager = DOWNLOAD_MANAGER.write().await;
    manager.pause_download(&download_id)
}

#[tauri::command]
pub async fn resume_download(download_id: String) -> Result<(), String> {
    let mut manager = DOWNLOAD_MANAGER.write().await;
    manager.resume_download(&download_id)
}

#[tauri::command]
pub async fn remove_download(download_id: String) -> Result<(), String> {
    let mut manager = DOWNLOAD_MANAGER.write().await;
    manager.remove_download(&download_id)
}

#[tauri::command]
pub async fn clear_completed_downloads() -> Result<(), String> {
    let mut manager = DOWNLOAD_MANAGER.write().await;
    manager.clear_completed_downloads();
    Ok(())
}

#[tauri::command]
pub async fn get_downloads() -> Result<Vec<Download>, String> {
    let manager = DOWNLOAD_MANAGER.read().await;
    Ok(manager.get_downloads().into_iter().cloned().collect())
}

#[tauri::command]
pub async fn get_active_downloads() -> Result<Vec<Download>, String> {
    let manager = DOWNLOAD_MANAGER.read().await;
    Ok(manager.get_active_downloads().into_iter().cloned().collect())
}

#[tauri::command]
pub async fn get_download_stats() -> Result<DownloadStats, String> {
    let manager = DOWNLOAD_MANAGER.read().await;
    Ok(manager.get_download_stats())
}

#[tauri::command]
pub async fn set_download_directory(path: String) -> Result<(), String> {
    let mut manager = DOWNLOAD_MANAGER.write().await;
    manager.set_download_directory(PathBuf::from(path))
}

#[tauri::command]
pub async fn get_download_progress(download_id: String) -> Result<Option<f64>, String> {
    let manager = DOWNLOAD_MANAGER.read().await;
    Ok(manager.get_download_progress(&download_id))
}

#[tauri::command]
pub async fn export_downloads() -> Result<String, String> {
    let manager = DOWNLOAD_MANAGER.read().await;
    manager.export_downloads()
}